## Sheetal Mahajan_20MAI0066

# Logistic Regression and Naïve Bayes Classifier

In [1]:
#Import the libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [4]:
#Read the dataset onto a variable

train=pd.read_csv("D:\\College\\sem2\\Data warehosing and mining\\Lab\\titanic_dataset.csv") #titanic dataset
train

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [5]:
#Fix the Predictor Variables
df=train[['Survived','Pclass','Sex','Age','Fare']]

In [6]:
#Change male to '1' and female to '0'
df["Sex"]=df["Sex"].apply(lambda sex:1 if sex=="male" else 0)

<ipython-input-6-04b246c879e4>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Sex"]=df["Sex"].apply(lambda sex:1 if sex=="male" else 0)


In [7]:
df['Sex'].value_counts()

1    577
0    314
Name: Sex, dtype: int64

In [8]:
# Handling the Missing Values -  Data Imputation
df.isna().sum()

Survived      0
Pclass        0
Sex           0
Age         177
Fare          0
dtype: int64

In [9]:
df['Age']=df['Age'].fillna(df['Age'].median())  #Outliners

<ipython-input-9-4995529c93a9>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Age']=df['Age'].fillna(df['Age'].median())  #Outliners


In [10]:
# Take a look at the Dataframe
df.head()

Survived  Pclass  Sex   Age     Fare
0         0       3    1  22.0   7.2500
1         1       1    0  38.0  71.2833
2         1       3    0  26.0   7.9250
3         1       1    0  35.0  53.1000
4         0       3    1  35.0   8.0500

In [11]:
# Set the Predictor(x) and Response(Y) variables
X=df.drop("Survived", axis=1)
Y=df["Survived"]

## 1. Logistic Regression: 

Logistic regression is a type of regression analysis. Regression analysis is a type of predictive modeling technique which is used to find the relationship between a dependent variable (usually known as the “Y” variable) and either one independent variable (the “X” variable) or a series of independent variables. When two or more independent variables are used to predict or explain the outcome of the dependent variable, this is known as multiple regression. Regression analysis can be broadly classified into two types: Linear regression and logistic regression.

Logistic regression is a classification algorithm. It is used to predict a binary outcome based on a set of independent variables.A binary outcome is one where there are only two possible scenarios—either the event happens (1) or it does not happen (0). Independent variables are those variables or factors which may influence the outcome (or dependent variable).

In [12]:
#Splitting into training and test set
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.25, random_state=42)

In [13]:
#Logistic regression Model
from sklearn.linear_model import LogisticRegression
logit = LogisticRegression()
logit.fit(X_train, Y_train)

LogisticRegression()

In [14]:
#Compute the Predictions 
Y_pred_L= logit.predict(X_test)
Y_pred_L

array([0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0,
       0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0,
       1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0,
       0, 1, 0], dtype=int64)

In [15]:
#Confusion Matrix
from sklearn.metrics import confusion_matrix
cm_L=confusion_matrix(Y_test,Y_pred_L)
cm_L

array([[111,  23],
       [ 25,  64]], dtype=int64)

In [16]:
confusion_L=pd.crosstab(Y_pred_L,Y_test,rownames=['predicted'],colnames=['actual'])
print("\n The Confusion Matrix is:")
confusion_L


 The Confusion Matrix is:


actual       0   1
predicted         
0          111  25
1           23  64

In [17]:
#Accuracy Score
from sklearn.metrics import accuracy_score
accuracy_logistics=accuracy_score(Y_test,Y_pred_L)
print("Accuracy using Logistic Regression is: ",accuracy_logistics)

Accuracy using Logistic Regression is:  0.7847533632286996


In [18]:
#Classification Report
from sklearn.metrics import classification_report
report=classification_report(Y_test,Y_pred_L)
print(report)

              precision    recall  f1-score   support

           0       0.82      0.83      0.82       134
           1       0.74      0.72      0.73        89

    accuracy                           0.78       223
   macro avg       0.78      0.77      0.77       223
weighted avg       0.78      0.78      0.78       223



## 2. Gaussian Naive Bayes Classifier

Naive Bayes classifiers are a collection of classification algorithms based on Bayes’ Theorem. It is not a single algorithm but a family of algorithms where all of them share a common principle, i.e. every pair of features being classified is independent of each other.

Naive Bayes is a simple technique for constructing classifiers: models that assign class labels to problem instances, represented as vectors of feature values, where the class labels are drawn from some finite set. There is not a single algorithm for training such classifiers, but a family of algorithms based on a common principle: all naive Bayes classifiers assume that the value of a particular feature is independent of the value of any other feature, given the class variable. For example, a fruit may be considered to be an apple if it is red, round, and about 10 cm in diameter. A naive Bayes classifier considers each of these features to contribute independently to the probability that this fruit is an apple, regardless of any possible correlations between the color, roundness, and diameter features.

In [19]:
#Gaussian Naive Bayes Classifier
from sklearn.naive_bayes import GaussianNB
gnbt=GaussianNB()
gnbt.fit(X_train, Y_train)

GaussianNB()

In [20]:
#Compute the Predictions
Y_pred_G= gnbt.predict(X_test)
Y_pred_G

array([0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0,
       0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0,
       1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0,
       0, 1, 0], dtype=int64)

In [21]:
#Confusion Matrix
from sklearn.metrics import confusion_matrix
cm_G=confusion_matrix(Y_test,Y_pred_G)
cm_G

array([[105,  29],
       [ 24,  65]], dtype=int64)

In [22]:
#Accuracy Score
from sklearn.metrics import accuracy_score
accuracy_gnbt=accuracy_score(Y_test,Y_pred_G)
print("Accuracy using Gaussian Naïve Bayes classifier is: ",accuracy_gnbt)

Accuracy using Gaussian Naïve Bayes classifier is:  0.7623318385650224


In [23]:
#Classification Report
from sklearn.metrics import classification_report
report_g=classification_report(Y_test,Y_pred_G)
print(report_g)

              precision    recall  f1-score   support

           0       0.81      0.78      0.80       134
           1       0.69      0.73      0.71        89

    accuracy                           0.76       223
   macro avg       0.75      0.76      0.75       223
weighted avg       0.77      0.76      0.76       223



### Achieved an accuracy using Logistic Regression as  0.78475 and accuracy for Gaussian Naïve Bayes classifier as  0.76233